Implement base functionality for kafka connector inserts

charlesjmorgan opened the pull request:
This is the first of 5 PRs based on #4230 

This PR focuses on adding code to the kafka connector to support inserts. Something to note is that inserting won't actually work because none of the encoders are included. Each of the four encoders will have their own follow-up PRs after this one gets merged. This PR also lays out the base interfaces and classes for the encoder.

I have changed how the encoder works since #4230 . Each `RowEncoder` has a collection field that will store row data until requested as a `byte[]` by the `KafkaPageSink`. The page sink will call overloaded `put(EncoderColumnHandle, value)` methods to add values to the row. Then once the page sink gets the `byte[]` from the `RowEncoder` it calls a `clear()` method that clears the backing collection.

This PR is part of the implementation for #3980 

**edit**: I added the CSV encoder to a separate commit in this pr just to have an encoder that works

**edit**: Other encoders are blocked by this pr for now. They can be found at these links
- Json Encoder PR - [https://github.com/charlesjmorgan/presto/pull/4](url)
- Raw Encoder PR - [https://github.com/charlesjmorgan/presto/pull/3](url)
- Avro Encoder PR - [https://github.com/charlesjmorgan/presto/pull/5](url)

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,156 @@
+        this.keyEncoder = requireNonNull(keyEncoder, "keyEncoder is null");
+        this.messageEncoder = requireNonNull(messageEncoder, "messageEncoder is null");
+        this.session = requireNonNull(session, "session is null");
+        this.producer = requireNonNull(producerFactory.create(new ByteArraySerializer(), new ByteArraySerializer()), "producerFactory is null");
```
Do we need to pass serializers here? could this be internal implementation detail of `KafkaProducerFactory`? You could change `KafkaProducerFactory` to always create `KafkaProducer<byte[], byte[]>` (no need for generics).

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,156 @@
+{
+    private final String topicName;
+    private final List<KafkaColumnHandle> columns;
+    private final RowEncoder keyEncoder;
```
it is not used here.

charlesjmorgan replied:
added to `ProducerMessage`

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -0,0 +1,113 @@
+        ImmutableSet.Builder<EncoderColumnHandle> messageColumns = ImmutableSet.builder();
+        handle.getColumns().forEach(col -> {
+            if (col.isInternal()) {
+                throw new PrestoException(GENERIC_USER_ERROR, "unexpected internal column");
```
Why is this user error?

charlesjmorgan replied:
Good point, I added a few new `KafkaErrorCodes` and changed that exception

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -0,0 +1,113 @@
+    public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorInsertTableHandle tableHandle)
+    {
+        requireNonNull(tableHandle, "tableHandle is null");
+        checkArgument(tableHandle instanceof KafkaTableHandle, "tableHandle is not an instance of KafkaTableHandle");
```
I think this check is not needed.

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPlugin.java:
```diff
@@ -25,6 +25,7 @@
 {
     public static final Module DEFAULT_EXTENSION = binder -> {
         binder.install(new KafkaConsumerModule());
+        binder.install(new KafkaProducerModule());
```
`PlainTextKafkaProducerFactory` is not extendable, there is no need to install this module here. Install it in `KafkaConnectorFactory` in a normal way.

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaProducerModule.java:
```diff
@@ -0,0 +1,39 @@
+            binder.bind(Properties.class).toInstance(new Properties());
+            binder.bind(PlainTextKafkaProducerFactory.class).toConstructor(PlainTextKafkaProducerFactory.class.getConstructor(KafkaConfig.class, Properties.class)).in(Scopes.SINGLETON);
+        }
+        catch (NoSuchMethodException e) {
```
Why do you need this `try-catch`?

charlesjmorgan replied:
`NoSuchMethodException` is thrown by `class.getConstructor()` if the constructor doesn't exist

losipiuk replied:
Why not just 
```java
        binder.bind(PlainTextKafkaProducerFactory.class).in(Scopes.SINGLETON);
```

charlesjmorgan replied:
Injection is new to me, didn't know I could do that lol

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaProducerModule.java:
```diff
@@ -0,0 +1,39 @@
+            binder.bind(PlainTextKafkaProducerFactory.class).toConstructor(PlainTextKafkaProducerFactory.class.getConstructor(KafkaConfig.class, Properties.class)).in(Scopes.SINGLETON);
+        }
+        catch (NoSuchMethodException e) {
+            throw new PrestoException(GENERIC_USER_ERROR, e);
```
Why is it an user error?

kokosing commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/EncoderErrorCode.java:
```diff
@@ -0,0 +1,40 @@
+
+    EncoderErrorCode(int code, ErrorType type)
+    {
+        errorCode = new ErrorCode(code + 0x0101_0000, name(), type);
```
make sure you are using unique code:

```
kokosing@m16:~/presto$ git grep '0x0101_0000'
presto-record-decoder/src/main/java/io/prestosql/decoder/DecoderErrorCode.java:        errorCode = new ErrorCode(code + 0x0101_0000, name(), type);
```

Why do you use `io.prestosql.plugin.kafka.KafkaErrorCode`?

charlesjmorgan replied:
I deleted `EncoderErrorCode` because I don't use the error code in this pr, I think it might be better to just add a `KAFKA_ENCODER_ERROR` to `KafkaErrorCode` in the future

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaQueryRunner.java:
```diff
@@ -148,7 +148,7 @@ private static DistributedQueryRunner createKafkaQueryRunner(
             Map<String, String> kafkaProperties = new HashMap<>(ImmutableMap.copyOf(extraKafkaProperties));
             kafkaProperties.putIfAbsent("kafka.nodes", testingKafka.getConnectString());
             kafkaProperties.putIfAbsent("kafka.connect-timeout", "120s");
-            kafkaProperties.putIfAbsent("kafka.default-schema", "default");
```
separate commit

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaQueryRunner.java:
```diff
@@ -203,6 +203,7 @@ public static void main(String[] args)
         Logging.initialize();
         DistributedQueryRunner queryRunner = builder(new TestingKafka())
                 .setTables(TpchTable.getTables())
+                .setExtraProperties(ImmutableMap.of("http-server.http.port", "8080"))
```
separate commit

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -44,18 +50,17 @@
 {
     private TestingKafka testingKafka;
-    private String topicName;
+    private String rawTopicName;
```
what is a raw topic name? how does it differ to topic name?

charlesjmorgan replied:
I can see how that might be confusing. It is meant to stand for the raw data format, I'll rename it.

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+    {
+        return testRoundTripAllFormatsData().stream()
+                .map(roundTripTestSetup -> new Object[] {roundTripTestSetup})
+                .toArray(Object[][]::new);
```
You might want to use `io.prestosql.testing.TestngUtils#toDataProvider`

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+    {
+        String tableName = roundTripTestSetup.getTableName();
+        String fieldNames = roundTripTestSetup.getFieldNames();
+        String fieldValues = roundTripTestSetup.getFieldValues();
```
I would inline these

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
     }
 
+    @Test(dataProvider = "testRoundTripAllFormatsDataProvider")
+    public void testRoundTripAllFormats(RoundTripTestSetup roundTripTestSetup)
```
roundTripTestSetup -> `testCase`?

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+                .build();
+    }
+
+    protected static final class RoundTripTestSetup
```
RoundTripTestSetup -> RoundTripTestCase?

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+                this.fieldValues = ImmutableList.copyOf(fieldValues);
+                this.length = fieldNames.size();
+            }
+            catch (IllegalArgumentException e) {
```
please remove this `try-catch`. `IllegalArgumentException` is already a runtime exception, also it is test code so there is no need to use `PrestoException`

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
 
+    @Test(dataProvider = "testRoundTripAllFormatsDataProvider")
+    public void testRoundTripAllFormats(RoundTripTestSetup roundTripTestSetup)
+    {
```
Don't you need to create a topic here?

charlesjmorgan replied:
I didn't include the actual test cases because the encoder formats aren't included in this pr. Maybe it would be best to not include this test in this pr either.

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+    }
+
+    protected static final class RoundTripTestSetup
+    {
```
The best would be have simple use case already implemented (possibly in separate commit) to see that all of this actually work.

charlesjmorgan replied:
I have one for each format when I open encoder prs

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+        }
+
+        private int getIndex(String fieldName)
+        {
```
fieldNames.indexOf(..)?

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+                checkArgument(index >= 0 && index < length, "index out of bounds");
+                return fieldNames.get(index);
+            }
+            catch (IllegalArgumentException e) {
```
remove try-catch

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+                return fieldNames.get(index);
+            }
+            else {
+                throw new PrestoException(GENERIC_USER_ERROR, format("field with name %s does not exist", fieldName));
```
use `checkArgument` instead

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+
+        public String getFieldNames()
+        {
+            StringBuilder sb = new StringBuilder();
```
`String.join()`?

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+                checkArgument(index >= 0 && index < length, "index out of bounds");
+                return fieldValues.get(index);
+            }
+            catch (IllegalArgumentException e) {
```
remove `try-catch`

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+                return fieldValues.get(index);
+            }
+            else {
+                throw new PrestoException(GENERIC_USER_ERROR, format("field with name %s does not exist", fieldName));
```
use `checkArgument` instead

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+
+        public String getFieldValues()
+        {
+            StringBuilder sb = new StringBuilder();
```
`String.join()`?

kokosing commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationSmokeTest.java:
```diff
@@ -116,6 +121,148 @@ private KafkaTopicFieldDescription createOneFieldDescription(String name, Type t
+        @Override
+        public String toString()
+        {
+            return tableName;  // for test case label in IDE
```
```suggestion
            return tableName; // for test case label in IDE
```

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/RowEncoder.java:
```diff
@@ -0,0 +1,115 @@
+
+    RowEncoder putNullValue(EncoderColumnHandle columnHandle);
+
+    default RowEncoder put(EncoderColumnHandle columnHandle, Object value)
```
please rename the `put` methods to `putObject`, `putLong`, `putInt` ...
Having that many overrides is very error prone. Especially when there are built-in type conversions in the language between the types for which you provide overrides.

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,156 @@
+            keyEncoder.clear();
+            messageEncoder.clear();
+        }
+        producer.flush();
```
`producer.send` returns a `Future`, not a `CompletableFuture`

charlesjmorgan replied:
is there a way to wake up the sender using the producer?

losipiuk replied:
Maybe sth like this would do?
```java
    private static class CountingSendCallback
            implements Callback
    {
        private final AtomicLong remaining;
        private final CompletableFuture<?> future;

        private CountingSendCallback(long pageSize) {
            this.remaining = new AtomicLong(pageSize);
            this.future = new CompletableFuture();
        }

        @Override
        public void onCompletion(RecordMetadata metadata, Exception exception)
        {
            // TODO possibly handle exception here
            remaining.decrementAndGet();
            completeFutureIfDone();
        }

        private void completeFutureIfDone()
        {
            if (remaining.get() <= 0) {
                future.complete(null);
            }
        }

        public CompletableFuture<?> getFuture()
        {
            return future;
        }
    }

    @Override
    public CompletableFuture<?> appendPage(Page page)
    {
        CountingSendCallback callback = new CountingSendCallback(page.getPositionCount());
        for (int position = 0; position < page.getPositionCount(); position++) {
   ....
            producer.send(new ProducerRecord<>(topicName, keyEncoder.toByteArray(), messageEncoder.toByteArray()), callback);
  .....
        }
        return callback.getFuture();
    }
```

losipiuk replied:
This also brings my attention to error handling. What would happen in current codebase if writing operation fails? Would that be propagated to INSERT query in presto. Maybe `producer.close();` would fail in such case but I am not sure. Do you know?

losipiuk replied:
> is there a way to wake up the sender using the producer?

Good point. Without it the code I pasted will not work. As the sending may basically not happen ever.
So maybe we need explicit `flush()` here or in `finish()`.
Does `producer.close()` we call in `finish()` also flush the sender queue. I skimmed Kafka code and it does not seem so.

It feels the safest for now to get back to what you had originally - flush after each page. And improve in followup PRs as we get more understanding.

charlesjmorgan replied:
that looks great, thank you! I think that you're right about handling exceptions in the `onCompletion` callback method. The docs for `KafkaProducer` say that `producer.close()` blocks until all previously sent requests complete.

losipiuk replied:
> that looks great

Not that much really :) - unless we can trigger actual sending to happen It is prone to starvation. As I wrote above. I think safest would be to do what you did originally to flush after each page or only in finish.

charlesjmorgan replied:
ok, we can come back to this later. Seems like it would still be smart to have a callback with an `onCompletion` method to handle any errors.

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/PlainTextKafkaProducerFactory.java:
```diff
@@ -0,0 +1,56 @@
+    private final Properties properties;
+
+    @Inject
+    public PlainTextKafkaProducerFactory(KafkaConfig kafkaConfig, Properties properties)
```
removed properties from constructor

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -0,0 +1,111 @@
+
+        RowEncoder keyEncoder = encoderFactory.create(
+                handle.getKeyDataFormat(),
+                ImmutableMap.of("dataSchema", getDataSchema(handle.getKeyDataSchemaLocation())),
```
Maybe we should have a constant for this?

charlesjmorgan replied:
if the encoder params are used just for the avro data schema, would it make sense to just pass an `Optional<String>` that stores the dataSchema? The map seems like overkill, and is kinda clunky, for just one variable.

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -0,0 +1,111 @@
+        catch (IOException e) {
+            throw new PrestoException(KAFKA_AVRO_SCHEMA_ERROR, format("Unable to read data schema at '%s'", dataSchemaLocation.get()), e);
+        }
+        return dataSchema;
```
It would seem more appropriate for this method to return an `Optional<String>`.

charlesjmorgan commented:
The RowEncoder now has a `putValueAt` method that has a block and position as parameters. This method handles getting the value from the block at the given position and calling the correct `put` method. I also changed the parameters of the `DispatchingRowEncoderFactory` and the `RowEncoderFactory` to no longer take a `Map` containing just the dataSchema. Instead it takes an `Optional<String>` that contains the parsed avro data schema if it exists. This `Map` was just used for the data schema so it seemed like overkill to me. I also added support for writing to multiple partitions, although this feature is not tested.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -95,17 +94,18 @@ public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHa
-        try {
-            if (dataSchemaLocation.isPresent()) {
-                dataSchema = Files.readString(Paths.get(dataSchemaLocation.get()));
+        if (dataSchemaLocation.isPresent()) {
```
```java
        return dataSchemaLocation.map(location -> {
            try {
                return Files.readString(Paths.get(location));
            }
            catch (IOException e) {
                throw new PrestoException(KAFKA_AVRO_SCHEMA_ERROR, format("Unable to read data schema at '%s'", location), e);
            }
        });
```

losipiuk commented:
> I also added support for writing to multiple partitions, although this feature is not tested.~

I would prefer to add it as a followup with test.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/RowEncoder.java:
```diff
@@ -29,7 +31,12 @@
     void clear();
 
-    RowEncoder putNullValue(EncoderColumnHandle columnHandle);
+    RowEncoder putValueAt(EncoderColumnHandle columnHandle, ConnectorSession session, Block block, int position);
```
Maybe `putValueFromBlock`? I do not know why but when I see `putValueAt` method which takes block I think it will be putting something into the block (though blocks are not mutable).

losipiuk replied:
Also can't you put default implementation already here?

charlesjmorgan replied:
right, that could definitely be confusing

charlesjmorgan replied:
I think so, would just be the same thing as what was in `KafkaPageSink#appendColumn`

losipiuk commented:
As you address the comments please squash the fixups. And a commit with simplest possible Encoder to this PR. It will make PR a bit larger, but he will have some validation that abstractions are sound.

charlesjmorgan commented:
> I would prefer to add it as a followup with test.

Ok, probably best not to rush things.

> And a commit with simplest possible Encoder to this PR.

They're all pretty equivalent in terms of complexity, maybe the csv one is simplest

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Override
+    public void abort() {}
```
We should probably close the producer here, as well.

aalbu commented:
> I would prefer to add it as a followup with test.

+1.

losipiuk commented:
> > And a commit with simplest possible Encoder to this PR.
> 
> They're all pretty equivalent in terms of complexity, maybe the csv one is simplest

CSV is fine

charlesjmorgan commented:
I've added a commit with the csv encoder

aalbu commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaQueryRunner.java:
```diff
@@ -134,9 +138,42 @@ private static DistributedQueryRunner createKafkaQueryRunner(
+                testingKafka.createTopics("write_test." + tableName);
+                SchemaTableName table = new SchemaTableName("write_test", tableName);
+                KafkaTopicDescription tableTemplate = topicDescriptionJsonCodec.fromJson(toByteArray(TestMinimalFunctionality.class.getResourceAsStream(format("/write_test/%s.json", tableName))));
+                Optional<KafkaTopicFieldGroup> key = tableTemplate.getKey();
+                if (tableTemplate.getKey().isPresent()) {
+                    if (tableTemplate.getKey().get().getDataSchema().isPresent()) {
+                        KafkaTopicFieldGroup keyTemplate = tableTemplate.getKey().get();
+                        key = Optional.of(new KafkaTopicFieldGroup(keyTemplate.getDataFormat(), Optional.of(KafkaQueryRunner.class.getResource(keyTemplate.getDataSchema().get()).getPath()), keyTemplate.getFields()));
+                    }
+                }
```
```suggestion
                Optional<KafkaTopicFieldGroup> key = tableTemplate.getKey()
                        .map(keyTemplate -> new KafkaTopicFieldGroup(
                                keyTemplate.getDataFormat(),
                                keyTemplate.getDataSchema().map(schema -> KafkaQueryRunner.class.getResource(schema).getPath()),
                                keyTemplate.getFields()));
```

aalbu commented on presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaQueryRunner.java:
```diff
@@ -134,9 +138,42 @@ private static DistributedQueryRunner createKafkaQueryRunner(
+                    }
+                }
+                Optional<KafkaTopicFieldGroup> message = tableTemplate.getMessage();
+                if (tableTemplate.getMessage().isPresent()) {
```
You could use functional style here, too.

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/RowEncoder.java:
```diff
@@ -0,0 +1,190 @@
+        throw new UnsupportedOperationException(format("Column '%s' does not support 'null' value", columnHandle.getName()));
+    }
+
+    default RowEncoder putObject(EncoderColumnHandle columnHandle, Object value)
```
no, I'll remove it

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/csv/CsvRowEncoder.java:
```diff
@@ -0,0 +1,200 @@
+                String columnName = columnHandle.getName();
+                Type columnType = columnHandle.getType();
+
+                checkArgument(!columnHandle.isInternal(), "unexpected internal column'%s'", columnName);
```
right, I noticed that internal columns don't ever seem to make it to the encoder

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/csv/CsvRowEncoder.java:
```diff
@@ -0,0 +1,200 @@
+                OutputStreamWriter outsWriter = new OutputStreamWriter(byteArrayOuts, StandardCharsets.UTF_8);
+                CSVWriter writer = new CSVWriter(outsWriter, ',', '"', "")) {
+            writer.writeNext(row);
+            writer.flush();
```
Even if I get the byte array before it's closed?

aalbu replied:
No, I missed the fact that the return is in the try-with-resources.

findepi commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/PlainTextKafkaProducerFactory.java:
```diff
@@ -0,0 +1,55 @@
+
+    public KafkaProducer<byte[], byte[]> create()
+    {
+        return new KafkaProducer<>(properties, new ByteArraySerializer(), new ByteArraySerializer());
```
`properties` is mutable, so defensive copy would be needed here

- change `Properties properties` field to `Map<...>  properties` and initialize it using `ImmutableMap`
- by using `Map<String, Object>` you can call the overload of `KafkaProducer` that does not require you to convert to `Properties`

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/RowEncoder.java:
```diff
@@ -0,0 +1,185 @@
+
+    void clear();
+
+    default RowEncoder putValueFromBlock(EncoderColumnHandle columnHandle, ConnectorSession session, Block block, int position)
```
The `position` argument here represents the index of the column from the table schema, which is not necessarily the same as the position in the result set, so we will still need a mapping (though it would be just an array lookup).

> Furthermore, we can guarantee that columns are being provided in order.
This would unlock certain optimizations in encoders like CSV or RAW, which
would be writing to a byte buffer directly.

I don't know if that's necessarily the case.  It would be pretty constraining for the API to require values be provided in the order columns appear in the table schema.

Perhaps we would be better off passing the entire page, instead of individual blocks, and let the encoder decide how to iterate over columns.

findepi replied:
> The `position` argument here represents the index of the column from the table schema

block+position represents a value (position is the row)

findepi replied:
> `void appendValue(Block block, int position)`

As discussed offline w @charlesjmorgan  -- `appendNextColumnValue`

> I don't know if that's necessarily the case. It would be pretty constraining for the API to require values be provided in the order columns appear in the table schema.

Why is that constraining? 
Wha would be the advantage of relaxing this?

aalbu replied:
> > The `position` argument here represents the index of the column from the table schema
> 
> block+position represents a value (position is the row)

Sorry, that was a dumb comment.

Where my head was at is that you might have a table schema like `foo (x varchar, y bigint, z boolean)` and you run `INSERT INTO foo (z, y) VALUES (true, 42)`, so if you don't specify the column, you need to perform a mapping for the column index from the result set schema to the table schema.

> Why is that constraining?

In the example above, you'd have to append `null`, then `42`, then `true`.  It seems like it requires a lot of work for the API client.

charlesjmorgan replied:
> this would allow encoder to precalculate ("compile") encoding strategies for each
of the columns

By this do you (@findepi) mean that the encoder could create a list of write functions for the different column types similar to the jdbc connector?

Are columns in pages always in the same order?

charlesjmorgan replied:
I'm worried that key columns and message columns being in the same page might mess up the ordering

charlesjmorgan replied:
It looks like pages always have the columns in the same (correct) order which is good. I could track the index of the key and message columns separately with two different ints.

>It seems like it requires a lot of work for the API client.

@aalbu I think that this is a good point. My only worry is that if we pass the entire page to the key/message encoder then the page includes both the key and message columns so it would need some way to distinguish these. Maybe we could pass a range of index values in the method too? Or would it be better to construct two `Block[]` for the key and message columns and then pass each of those once to their respective encoders to minimize the number of api calls needed.

charlesjmorgan replied:
also I'm not sure if the name `append...` really fits if we specify the position to put the value at

aalbu replied:
> It looks like pages always have the columns in the same (correct) order which is good

Ah, my incorrect assumption was that only the columns specified in the insert are going to be present on the page.  That makes my concern irrelevant.

findepi replied:
> By this do you (@findepi) mean that the encoder could create a list of write functions for the different column types similar to the jdbc connector?

it can, if it wishes to do so

i am not saying "do it". I am just saying "it is possible"

> Are columns in pages always in the same order?

yes

findepi replied:
> @aalbu I think that this is a good point. My only worry is that if we pass the entire page to the key/message encoder then the page includes both the key and message columns so it would need some way to distinguish these.

That's a very good point, which actually rules out passing full `Page` (+ position)

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,102 @@
+            PlainTextKafkaProducerFactory producerFactory)
+    {
+        this.topicName = requireNonNull(topicName, "topicName is null");
+        this.columns = requireNonNull(ImmutableList.copyOf(columns), "columns is null");
```
I think we should do some additional validation on the columns here.  For example, if the table schema specifies a key, we would need to ensure that all key columns are present.  We could also do the check for internal columns here - make sure that only supported ones are present.

findepi replied:
All columns are present by means of `io.prestosql.spi.connector.ConnectorMetadata#supportsMissingColumnsOnInsert`

aalbu replied:
Thanks.

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaMetadata.java:
```diff
@@ -107,13 +115,16 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
+    }
 
-        SchemaTableName schemaTableName = kafkaTableHandle.toSchemaTableName();
+    @SuppressWarnings("ValueOfIncrementOrDecrementUsed")
```
I need to get the columns using just the schema table name so I split the function into two parts. Would it be a good idea to give them different names? like `getColumnHandlesHelper` or something? The suppress warnings annotation was there originally.

losipiuk replied:
Naming is fine. As for the annotation,  drop it (and one below) in separate commit.

charlesjmorgan commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaErrorCode.java:
```diff
@@ -18,11 +18,16 @@
-    KAFKA_SPLIT_ERROR(0, EXTERNAL);
+    KAFKA_SPLIT_ERROR(0, EXTERNAL),
+    KAFKA_AVRO_SCHEMA_ERROR(1, EXTERNAL),
+    KAFKA_PRODUCER_ERROR(2, INTERNAL_ERROR),
+    KAFKA_ENCODER_ERROR(3, EXTERNAL)
```
Would it be better to just throw some sort of runtime exception in these cases or have a catch all `KAFKA_ERROR`?

charlesjmorgan commented:
Made changes based on the suggestions made by @findepi 

**Kafka inserts commit**
- Removed default implementations of methods in `RowEncoder` interface
- Added `AbstractRowEncoder` class that provides structure for the different data format encoder implementations
- Changed `putValue` methods to `appendValue` methods in the encoders that now only take the block and position as parameters
- `AbstractRowEncoder` now tracks the current column index for appending values to the row encoder, no longer have to pass column handle as parameter when appending
- Encoders now get the current session during implementation
- Changed name of field `keyDecoder` in `KafkaColumnHandle` to `keyCodec` so that it can be used without confusion for the decoder and the encoder
- Added check in `KafkaMetadata` to make sure insert columns match actual columns
- `KafkaPageSink#appendPage` now has some (limited) error handling if a producer record fails to send
- Changed column handles stored in row encoder from `Set` to `List` to preserve correct order
- The producer factory now uses an `ImmutableMap` to store config info

**CSV encoder commit**
- CSV encoder now uses `currentColumnIndex` to append values to the `String[]`
- CSV encoder uses `AbstractRowEncoder` constructor
- `kafka_key` column in csv table is no longer hidden, users are able to specify a key

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaErrorCode.java:
```diff
@@ -24,7 +24,7 @@
 {
     KAFKA_SPLIT_ERROR(0, EXTERNAL),
-    KAFKA_AVRO_SCHEMA_ERROR(1, EXTERNAL),
+    KAFKA_SCHEMA_ERROR(1, EXTERNAL),
```
This should go to previous commit.

charlesjmorgan replied:
right, thank you

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,99 @@
+                }
+            }
+            producer.send(new ProducerRecord<>(topicName, keyEncoder.toByteArray(), messageEncoder.toByteArray()), (recordMetadata, e) -> {
+                if (e != null) {
```
The callback can be run in very different thread than the one calling `producer.send`. Do we know if exception thrown by the callback will be always propagated back to Presto thread. I.e if we do not get exception directly here, will we still get it on `producer.flush()` called from `finish()`?

charlesjmorgan replied:
the callback is executed in the I/O thread of the producer. It doesn't look like flush throws an exceptions other than `InterruptException`

losipiuk replied:
Indeed it looks so. If the callback throws exception in later stage of sending it will be caught by try-catch in `ProducerBatch.completeFutureAndFireCallbacks` and just logged. See quoted method body below.
```java
    private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime, RuntimeException exception) {
        // Set the future before invoking the callbacks as we rely on its state for the `onCompletion` call
        produceFuture.set(baseOffset, logAppendTime, exception);

        // execute callbacks
        for (Thunk thunk : thunks) {
            try {
                if (exception == null) {
                    RecordMetadata metadata = thunk.future.value();
                    if (thunk.callback != null)
                        thunk.callback.onCompletion(metadata, null);
                } else {
                    if (thunk.callback != null)
                        thunk.callback.onCompletion(null, exception);
                }
            } catch (Exception e) {
                log.error("Error executing user-provided callback on message for topic-partition '{}'", topicPartition, e);
            }
        }

        produceFuture.done();
    }
```

Given such logic I think that instead of throwing exception in from our callback, we should just record that message sending failure happened. A simple counter, how many message-sending failures we observed should do. 
Then in `KafkaPageSink.finish()`, after `producer.close()`, we can introspect the counter, and if it is >0 we can throw `PrestoException` there

charlesjmorgan replied:
that sounds good!

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/AbstractRowEncoder.java:
```diff
@@ -0,0 +1,218 @@
+    /**
+     * Performs any checks on the column handle field values
+     */
+    protected abstract List<EncoderColumnHandle> validateColumns(List<EncoderColumnHandle> columnHandles);
```
It cannot return different list than one we pass to it as an argument, right? Otherwise the `columHandles` stored here would not be compatible with loop which calls to `appendColumnValue`. Change to void (unless I miss something).

charlesjmorgan replied:
It only returns a different list as a defensive copy, I don't think that this should affect how it functions

losipiuk replied:
I think we should do the defensive copy in `AbstractRowEncoder` and make `validate` return void so it is obvious it cannot change the list.

charlesjmorgan replied:
right, would be best to have `validateColumns` only validate the columns rather than validate-and-copy the columns

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/AbstractRowEncoder.java:
```diff
@@ -0,0 +1,218 @@
+    @Override
+    public RowEncoder appendColumnValue(Block block, int position)
+    {
+        Type type = columnHandles.get(currentColumnIndex).getType();
```
`checkState` that currentColumnIndex < columnHandles.size()

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/AbstractRowEncoder.java:
```diff
@@ -0,0 +1,218 @@
+    @Override
+    public void reset()
+    {
+        currentColumnIndex = 0;
```
should we `checkState` that `currentColumnIndex == columnHandles.size()`?

charlesjmorgan replied:
yes, that would be a good idea

charlesjmorgan replied:
I actually think this belongs in `toByteArray`, not in `reset`. So that if not all columns are appended previous row values can't be duplicated.

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/csv/CsvRowEncoder.java:
```diff
@@ -0,0 +1,151 @@
+    @Override
+    protected void clear()
+    {
+        row = new String[size];
```
not really needed right?

charlesjmorgan replied:
If we make sure that every index is updated for every row then no it is not needed

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/csv/CsvRowEncoder.java:
```diff
@@ -0,0 +1,151 @@
+    {
+        try (ByteArrayOutputStream byteArrayOuts = new ByteArrayOutputStream();
+                OutputStreamWriter outsWriter = new OutputStreamWriter(byteArrayOuts, StandardCharsets.UTF_8);
+                CSVWriter writer = new CSVWriter(outsWriter, ',', '"', "")) {
```
do we support overriding special characters on read? If so we should also support them on write.

charlesjmorgan replied:
You mean like using something other than a comma to separate the values?

losipiuk replied:
Yep. Or different quote char.

charlesjmorgan replied:
I don't see anything about that in the docs for read so I'm gonna say no it doesn't. Would be a nice feature tho

charlesjmorgan commented:
Changes based on review from @losipiuk 

**Kafka inserts commit**
- `validateColumns` returns void and `columnHandles` are copied in `AbstractRowEncoder` constructor
- Add check in `appendColumnValue` to make sure `currentColumnIndex` is not greater than number of columns
- Add callback in `KafkaPageSink` that counts number of producer records that fail to send and throw error if number is greater than 0

**CSV encoder commit**
- remove `size` field from `CsvRowEncoder`
- clear no longer does anything in the CSV encoder because there are checks to make sure that every column is appended for every new row
- `toByteArray` checks to make sure that every column has been updated with values for each new row

Rebased and brought in `KafkaColumnHandle` changes

losipiuk commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java:
```diff
@@ -0,0 +1,123 @@
+        this.messageEncoder = requireNonNull(messageEncoder, "messageEncoder is null");
+        requireNonNull(producerFactory, "producerFactory is null");
+        this.producer = producerFactory.create();
+        this.errorCounter = new ErrorCountingCallback();
```
Any idea if we can write a test which triggers errorcounter?

charlesjmorgan replied:
hmm not that I can think of, I'm not sure how to intentionally make a producer fail to send a record.

charlesjmorgan commented:
I'm feeling good about where this PR is at @losipiuk @aalbu any more comments?

aalbu commented on presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/RowEncoder.java:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * Resets the encoder so that it can be used for the next row
+     */
+    void reset();
```
I wonder if we need this method at all.  Perhaps the contract could be: add data, then generate the record, rinse, repeat.  Generating the `byte[]` could reset things (that would be an implementation detail).

charlesjmorgan replied:
I was thinking a similar thing. I just went through and updated all the encoders with the changes I've made and I realized that the `clear` method  was only used once by the `RawRowEncoder`.

charlesjmorgan commented:
The other encoders are blocked by this pr for now. They can be found here:
- Json Encoder PR - [https://github.com/charlesjmorgan/presto/pull/4](url)
- Raw Encoder PR - [https://github.com/charlesjmorgan/presto/pull/3](url)
- Avro Encoder PR - [https://github.com/charlesjmorgan/presto/pull/5](url)
@losipiuk @aalbu

losipiuk commented:
Thanks !
